README.md
# VXFLD

## VXLAN BUM Flooding Suite

VXFLD is a suite of tools that provides the ability to do VXLAN
BUM flooding using unicast instead of the traditional multicast.

This is accomplished using 2 components, [vxsnd](https://github.com/CumulusNetworks/vxfld/blob/master/vxsnd.rst)
and [vxrd](https://github.com/CumulusNetworks/vxfld/blob/master/vxrd.rst).

vxsnd provides the unicast BUM packet flooding and VTEP learing
capabilities while vxrd is a simple registration daemon designed to
register local VTEPs with a remote vxsnd daemon.

## TODO
- Requesting complete VNI dump from other node (initial startup)
- concurrency model
  - coroutines, eventlet/greenlet/greenthreads
  - 3 basic core functions today should have distince concurrency
    - BUM packet flooding
    - VNI registration
    - Replication
      - Sending VNI data to other node
      - Responding to VNI dump request from peer
- Unit tests
- pep8
- ability to run as non-priveleged user
- tiered replication/flooding


debian/changelog
vxfld (1.0-1) unstable; urgency=low

  * Initial release

 -- Nolan Leake <nolan@cumulusnetworks.com>  Fri, 11 Jul 2014 17:45:00 +0000



vxfld/common.py
#! /usr/bin/python
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2014 Cumulus Networks, Inc. All rights reserved.
# Copyright (C) 2014 Metacloud Inc.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc.
# 51 Franklin Street, Fifth Floor
# Boston, MA  02110-1301, USA.

""" Stuff that is common to both vxfld daemons. """
import os
import sys
import signal
import subprocess
import atexit
import argparse
import logging
import logging.handlers
import vxfld.config


def run_cmd(cmd, verbose=False):
    """
    Runs a command with Popen gathering output & status. Takes a str arg.
    Should take a list too
    """

    cmd_list = cmd.split()
    try:
        proc = subprocess.Popen(cmd_list, stdout=subprocess.PIPE,
                                stderr=subprocess.PIPE,
                                shell=False)
        stdout, stderr = proc.communicate()
        proc.wait()
    except Exception as e:
        raise RuntimeError('Failed to run shell cmd %s: %s' % (cmd, str(e)))
    return (stdout, stderr, proc.returncode)


def logger_setup(conf):
    """ Setup logging. """
    global lgr
    lgr = logging.getLogger('vxfld')

    p = os.path.basename(sys.argv[0])
    lgr_fmt = '%%(asctime)s %s %%(levelname)s: %%(message)s' % p

    if conf.logdest == 'syslog':
        h = logging.handlers.SysLogHandler(address='/dev/log')
        syslog_fmt = '%s: %%(levelname)s: %%(message)s' % p
        f = logging.Formatter(fmt=syslog_fmt)
        h.setFormatter(f)
        lgr.addHandler(h)
        lgr.setLevel(conf.loglevel)
        return lgr
    if conf.logdest == 'stdout':
        logging.basicConfig(level=conf.loglevel,
                            format=lgr_fmt,
                            datefmt='%H:%M:%S')
        return lgr

    # logdest is a file
    logging.basicConfig(filename=conf.logdest,
                        level=conf.loglevel,
                        format=lgr_fmt,
                        datefmt='%H:%M:%S')
    return lgr


def write_pidfile(pf):
    """ check/create/delete pidfile to avoid duplicate instances

    pidfile is a global since we have to register it's delete-on-exit handler
    pid fd is global. Otherwise OS reclaims lock if fd closed
    """

    import fcntl

    global pidfile
    global pidfd
    pidfile = os.path.abspath(pf)

    pid = str(os.getpid())
    try:
        # Mode is append so we don't truncate until we get lock
        f = open(pidfile, 'a')
    except:
        raise RuntimeError("Cannot open pid file %s" % pidfile)
    lgr.debug("Opened pid file %s" % pidfile)
    try:
        fcntl.lockf(f, fcntl.LOCK_EX | fcntl.LOCK_NB)
    except Exception:
        raise RuntimeError("Unable to lock pid file")

    # Now we can register the callback to del the file on exit and
    # write the pid into the file
    atexit.register(delpid)
    f.truncate(0)
    f.write("{0}\n".format(pid))
    f.flush()
    pidfd = f

    # we're good to go
    #
    return pid


def delpid():
    try:
        os.remove(pidfile)
    except Exception as e:
        lgr.critical("Unable to remove pid file on exit: %s" % str(e))


def start_monit(service):
    # On failure just log.  Don't pass exception up
    try:
        (out, err, errno) = run_cmd('monit monitor %s 2>/dev/null' % service)
        if errno:
            raise RuntimeError('monit error %d', errno)
    except:
        lgr.warning('Failed to initialize monitoring with monit')


def term_handler(signum, frame):
    sys.exit(0)


def common_parser():
    """ Argparser for common cmd line args. """

    prsr = argparse.ArgumentParser()
    prsr.add_argument('-c', '--config-file',
                      default='/etc/vxfld.conf',
                      help='The config file to read in at startup')
    prsr.add_argument('-d', '--daemon',
                      action='store_true',
                      help='Run as a daemon program')
    prsr.add_argument('-p', '--pidfile',
                      help='File to write the process ID')
    prsr.add_argument('-l', '--logdest',
                      help='The destination for log records.  May be '
                           '"stdout", "syslog" or a file name')
    prsr.add_argument('-L', '--loglevel',
                      help='The severity level for which log records are '
                           'written to the log.  Allowed values in porder of '
                           'severity are: debug, info, warning, error, '
                           'critical')

    return prsr


def initial_setup(args):
    """ All the initial setup common to both daemons. """

    # Init config
    try:
        conf = vxfld.config.init(args)
    except Exception as e:
        # On error expecting a RuntimeError with nice msg
        sys.stderr.write('%s\n' % str(e))
        sys.exit(1)

    # Set up signal handlers before daemonizing
    signal.signal(signal.SIGINT, term_handler)
    signal.signal(signal.SIGTERM, term_handler)
    signal.signal(signal.SIGHUP, term_handler)

    if conf.daemon:
        import daemon

        dc = daemon.DaemonContext(working_directory='/var/run')
        dc.open()

    # set up logging.  Must be after daemonizing.
    try:
        lgr = logger_setup(conf)
    except Exception as e:
        # If daemonized, stderr is /dev/null!
        sys.stderr.write('Unable to set up logging: %s\n' % str(e))
        sys.exit(1)

    # Announce my existence
    lgr.info("Starting (pid %d) ..." % os.getpid())
    atexit.register(lgr.info, "Terminating (pid %d)" % os.getpid())

    # Make sure that I am not already running
    try:
        write_pidfile(conf.pidfile)
    except RuntimeError as e:
        lgr.critical(str(e))
        sys.exit(1)

    return conf, lgr


vxfld/__init__.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2014 Cumulus Networks, Inc. All rights reserved.
# Copyright (C) 2014 Metacloud Inc.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc.
# 51 Franklin Street, Fifth Floor
# Boston, MA  02110-1301, USA.


vxfld/vxfldpkt.py
#! /usr/bin/python
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2014 Cumulus Networks, Inc. All rights reserved.
# Copyright (C) 2014 Metacloud Inc.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc.
# 51 Franklin Street, Fifth Floor
# Boston, MA  02110-1301, USA.

import dpkt
import struct
import socket

# Message types

version = 1  # Protocol version


class MsgType():
    unknown = 0  # Never used
    refresh = 1
    resend = 2


class PktError(Exception):
    def __init__(self, msg):
        self.msg = msg


class Refresh(dpkt.Packet):
    """Packet sent between vxsnd entities to refresh the vvtuples."""

    __hdr__ = (
        ('version', 'B', 0x01),  # version of the protocol Packet
        ('type', 'B', MsgType.refresh),
        ('originator', 'H', 0),  # should be all flags
        ('holdtime', 'H', 0)
    )

    def __init__(self, *args, **kwargs):
        dpkt.Packet.__init__(self, **kwargs)
        self.vni_vteps = dict()
        if args:
            self.unpack(args[0])

    def unpack(self, buf):
        dpkt.Packet.unpack(self, buf)
        if self.version != version:
            raise PktError("Wrong version")
        data = self.data
        pos = 0
        data_len = len(data)
        while pos < data_len:
            (vni, cnt) = struct.unpack('>IH', data[pos:pos + 6])
            pos += 6
            if pos + cnt * 4 > data_len:
                raise PktError("Short packet")
            if not self.vni_vteps.get(vni, None):
                self.vni_vteps[vni] = []
            while cnt:
                ip = data[pos:pos + 4]
                self.vni_vteps[vni].append(socket.inet_ntoa(ip))
                cnt -= 1
                pos += 4

    def __str__(self):
        s = ''
        for (vni, iplist) in self.vni_vteps.items():
            s += struct.pack('>IH', vni, len(iplist))
            for ip in iplist:
                s += socket.inet_aton(ip)
        return self.pack_hdr() + s

    def __len__(self):
        cnt = 0
        for (vni, iplist) in self.vni_vteps.items():
            cnt += 4 + 2 + 4 * len(iplist)
        return self.__hdr_len__ + cnt

    def add_vni_vteps(self, vp):
        '''Add a set of <vni vtep_list>.'''

        for (vni, l) in vp.items():
            iplist = self.vni_vteps.get(vni, None)
            if iplist:
                iplist.extend(l)
            else:
                self.vni_vteps[vni] = l


vxfld/config.py
#! /usr/bin/python
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2014 Cumulus Networks, Inc. All rights reserved.
# Copyright (C) 2014 Metacloud Inc.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc.
# 51 Franklin Street, Fifth Floor
# Boston, MA  02110-1301, USA.

"""
Configuration object

Typical usage:
    Create object (config = Config()
    load in checker functions
    load with defaults
    load with cmd line options from argparse, if desired
"""

import socket


class Config:
    """ Generic Config class. """

    def __init__(self):
        self._checkers = {}

    def set_param(self, name, value):
        """
        If there is a checker function for this parameter, call it
        with value.  Otherwise simply set an attribute with the name
        to the value.
        """

        if name in self._checkers:
            # Let checker do it's thing before setting
            result = self._checkers[name][0](self, value)
            setattr(self, name, result)

            # Now call cb function.
            if self._checkers[name][1]:
                self._checkers[name][1](self)
        else:
            # No checker, just set val
            setattr(self, name, value)

    def del_param(self, name):
        """ Delete the parameter altogether.

        Useful for clearing out a parameter where the set is additive.
        """
        pass

    def read_file(self, fn):
        """ Read config from file fn and process each line. """

        COMMENT_CHAR = '#'
        OPTION_CHAR = '='
        try:
            fd = open(fn)
        except Exception as e:
            raise RuntimeError('Cannot open config file %s: %s' % (fn, str(e)))
        with fd:
            for line in fd:
                # First, remove comments:
                if COMMENT_CHAR in line:
                    line, comment = line.split(COMMENT_CHAR, 1)
                # Second, find lines with an option=value:
                if OPTION_CHAR in line:
                    option, value = line.split(OPTION_CHAR, 1)
                    option = option.strip()
                    value = value.strip()
                    if hasattr(self, option):
                        self.set_param(option.strip(), value.strip())
                    else:
                        msg = 'Unknown variable "%s" in config file %s' % \
                              (option, fn)
                        raise RuntimeError(msg)

    def checker(self, func, param=None, callback=None):
        """
        Set the function as the checker function for the named
        parameter.  If no parameter name given, use the function name.
        """

        if not param:
            param = func.__name__
        self._checkers[param] = (func, callback)

    # Common checker functions
    def int_checker(self, param, cb=None):
        self._checkers[param] = (Config._int_checker, cb)

    def _int_checker(self, val):
        """ Returns int from string or int. """
        try:
            return int(val)
        except:
            raise RuntimeError('Invalid integer value %s' % val)

    def bool_checker(self, param, cb=None):
        self._checkers[param] = (Config._bool_checker, cb)

    def _bool_checker(self, val):
        """ Returns True for string true or boolean True. """

        if type(val) is bool:
            return val
        val = val.lower()
        if val == 'true':
            return True
        return False

    def list_checker(self, param, cb=None):
        self._checkers[param] = (Config._list_checker, cb)

    def _list_checker(self, val):
        """ Returns a list by splitting string. """
        try:
            return val.split()
        except:
            raise RuntimeError('Invalid string for list value %s' % val)

    def addr_checker(self, param, cb=None):
        self._checkers[param] = (Config._addr_checker, cb)

    def _addr_checker(self, s):
        """ Returns the result of gethostbyname on a string.

        verifies correctness of a dotted decimal specificationi.
        """

        try:
            return socket.gethostbyname(s)
        except:
            raise RuntimeError('Invalid address %s' % s)


defaults = {
    # A dictionary to prime the config object with all config attrs plus
    # some constants.
    #
    # This dict sets up all the default values.  It also primes the
    # config obj with all the attrs needed so we don't have to test
    # for existence every time a param is accessed.  Also, loading
    # from a config file requires the config variables to be
    # pre-loaded since a non-existent attribute is taken to be a
    # config file error.

    # Common to both snd and rd
    'protocol_version': '0.1',  # a constant
    'loglevel': 'INFO',
    'logdest': 'syslog',
    'monitor': 'false',  # set to true to monitor with monit
    'pidfile': '/var/run/vxfld.pid',
    'vxlan_port': '4789',  # port for vxlan tunnel pkts
    'vxfld_port': '10001',  # port for vxfld messages
    'holdtime': '90',  # how long to hold soft state

    #  vxsnd specific.  Add these here to prime the config object with
    #  these attributes before config file is read.  Does no harm if
    #  included by vxrd
    'address': '',
    'install_addr': 'false',
    'servers': '',
    'age_check': '90',  # frequency to age out stale fdb entries
    'vtep_membership': '',  # not yet implemented
    'max_packet_size': '1500',
    'receive_queue': '131072',

    #  .. and vxrd specific.
    'local_addr': '',  # Used if none configured on vxlan if
    'svcnode': '',  # Used if none configured on vxlan if
    'refresh_rate': '3',  # how often to refresh within holdtime
    'config_check_rate': 30  # secs between checking for config changes
}


def loglevel(c, val):
    """
    Checker functions for attrs which go beyond the basic int, list, bool.
    """

    import logging

    lvl = getattr(logging, val.upper(), None)
    if not lvl:
        raise RuntimeError('Invalid log level %s' % val)
    logging.getLogger('vxfld').setLevel(lvl)
    return lvl


def servers(c, val):
    # split the string and resolve to set of addresses.
    result = set()
    l = val.split()
    for s in l:
        try:
            a = socket.gethostbyname(s)
        except:
            lgr.warning('Cannot resolve address for server %s' % s)
            continue
        result |= set([a])
    return result


def init(args):
    """ Called to setup initial Config object. """
    config = Config()

    # set up checker funtions first

    # common
    config.checker(loglevel)
    config.bool_checker('monitor')
    config.int_checker('vxlan_port')
    config.int_checker('vxfld_port')
    config.int_checker('holdtime')

    # vxsnd
    config.addr_checker('address')
    config.bool_checker('install_addr')
    config.checker(servers)
    config.int_checker('age_check')
    config.int_checker('max_packet_size')
    config.int_checker('receive_queue')

    # vxrd
    config.addr_checker('local_addr')
    config.addr_checker('svcnode')
    config.int_checker('refresh_rate')
    config.int_checker('config_check_rate')

    # Load the defaults
    for (p, v) in defaults.items():
        config.set_param(p, v)

    # Over-write with settings from config file
    #  args must contain this attr
    config.read_file(args.config_file)

    for (p, v) in vars(args).items():
        if v is not None:
            config.set_param(p, v)

    return config


if __name__ == '__main__':
    # This main is currently broken
    #

    import logging

    config = init()

    lgr = logging.getLogger('vxfld')
    print 'Default log level: ', lgr.getEffectiveLevel()
    config.set_param('loglevel', 'DEBUG')
    print 'Changed log level: ', lgr.getEffectiveLevel()

    print 'Default vxlan port:', config.vxlan_port
    config.set_param('vxlan_port', '8081')
    print 'Changed vxlan port:', config.vxlan_port
    print vars(config)
